### PR TITLE
More tests for sumi

### PR DESCRIPTION
**Pull request by azaidy**
This PR adds testbenches and tests for umi_mux, umi_demux, and umi_switch.
Additionally, it converts umi_ram and umi_regif to pytest.

### CODE REVIEW

File: umi/sumi/tests/test_demux.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        print(f"Sending #{count} cmd: 0x{txp.cmd:08x} srcaddr: 0x{srcaddr:08x} "
+              f"dstaddr: 0x{dstaddr:08x} to port {dstport}")
+        # send the packet to both simulation and local queues
+        umi_q.send(txp)
+        tee[dstport].send(txp)
+
+
+def test_demux(sumi_dut, random_seed, sb_umi_valid_mode, sb_umi_ready_mode):
+    n = 100
+    ports = 4
```
Comment (gadfort) [inline umi/sumi/tests/test_demux.py:32]: what are these magic numbers?

File: umi/sumi/tests/test_mux.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        print(f"Sending #{count} cmd: 0x{txp.cmd:08x} srcaddr: 0x{srcaddr:08x} "
+              f"dstaddr: 0x{dstaddr:08x}")
+        # send the packet to both simulation and local queues
+        umi_q.send(txp)
+        tee.send(txp)
+
+
+def test_mux(sumi_dut, random_seed, sb_umi_valid_mode, sb_umi_ready_mode):
+    n = 100
+    ports = 4
```
Comment (gadfort) [inline umi/sumi/tests/test_mux.py:31]: same

File: umi/sumi/tests/test_demux.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    for count in range(n*ports):
+        dstport = np.random.randint(0, ports)
+        dstaddr = (2**8)*np.random.randint(0, 2**32) + (2**(40+dstport))
+        srcaddr = (2**8)*np.random.randint(0, 2**32) + (2**(40+ports))
+        txp = random_umi_packet(dstaddr=dstaddr, srcaddr=srcaddr)
+        print(f"Sending #{count} cmd: 0x{txp.cmd:08x} srcaddr: 0x{srcaddr:08x} "
+              f"dstaddr: 0x{dstaddr:08x} to port {dstport}")
+        # send the packet to both simulation and local queues
+        umi_q.send(txp)
+        tee[dstport].send(txp)
```
Comment (gadfort) [inline umi/sumi/tests/test_demux.py:27]: I feel like this code appears a lot...

File: umi/sumi/tests/test_regif.py [hunk 4]
```diff
@@ -80,34 +52,35 @@ def apply_atomic(origdata, atomicdata, operation, maxrange):
...
 
 
-def main(vldmode="2", rdymode="2", n=100, host2dut="host2dut_0.q", dut2host="dut2host_0.q"):
-    # clean up old queues if present
-    for q in [host2dut, dut2host]:
-        delete_queue(q)
+def test_regif(sumi_dut, random_seed, sb_umi_valid_mode, sb_umi_ready_mode):
 
-    verilator_bin = build_testbench()
+    n = 100
```
Comment (gadfort) [inline umi/sumi/tests/test_regif.py:57]: magic number

**Review by gadfort — CHANGES_REQUESTED** [review]
overall looks good, but I'm not a fan of magic numbers, maybe you can add some comments around them?

File: umi/sumi/tests/test_demux.py (hunk 1 above)
Comment (azaidy) [reply umi/sumi/tests/test_demux.py]: Added comments to explain them. n is just the number of transaction. ports must the number of output ports in the instantiation from the testbench

File: umi/sumi/tests/test_mux.py (hunk 2 above)
Comment (azaidy) [reply umi/sumi/tests/test_mux.py]: Done.

File: umi/sumi/tests/test_regif.py (hunk 4 above)
Comment (azaidy) [reply umi/sumi/tests/test_regif.py]: Done.

File: umi/sumi/tests/test_demux.py (hunk 3 above)
Comment (azaidy) [reply umi/sumi/tests/test_demux.py]: I refactored the tests and pulled this code out into a common file. Variants of it appeared in demux, mux, and switch. In order to pull the code out, the variants had to be merged into a common format which meant the tests changed slightly as well.

**Review by gadfort — APPROVED** [review]

**Pull request merged**